server/tilesource/base.py

```diff
@@ -352,29 +576,217 @@ def getRegion(self, width=None, height=None, **kwargs):
             # PIL will reallocate buffers that aren't in 'raw', RGBA, 0, 1.
             # See PIL documentation and code for more details.
             b'\x00' * (regionWidth * regionHeight * 4), 'raw', 'RGBA', 0, 1)
-        for x in range(xmin, xmax):
-            for y in range(ymin, ymax):
-                tileData = self.getTile(
-                    x, y, preferredLevel, pilImageAllowed=True,
-                    sparseFallback=True)
-                if not isinstance(tileData, PIL.Image.Image):
-                    tileData = PIL.Image.open(BytesIO(tileData))
-                posX = x * metadata['tileWidth'] - left
-                posY = y * metadata['tileHeight'] - top
-                # Add each tile to the image.  PIL crops these if they are off
-                # the edge.
-                image.paste(tileData, (posX, posY),
-                            tileData if mode == 'RGBA' and
-                            tileData.mode == 'RGBA' else None)
-
+        for tile in self._tileIterator(iterInfo):
+            # Add each tile to the image.  PIL crops these if they are off the
+            # edge.
+            image.paste(tile['tile'], (tile['x'] - left, tile['y'] - top),
+                        tile['tile'] if mode == 'RGBA' and
+                        tile['tile'].mode == 'RGBA' else None)
         # Scale if we need to
+        width = int(math.floor(width))
+        height = int(math.floor(height))
         if width != regionWidth or height != regionHeight:
             image = image.resize(
                 (width, height),
                 PIL.Image.BICUBIC if width > regionWidth else
                 PIL.Image.LANCZOS)
         return self._encodeImage(image, **kwargs)
 
+    def getMagnification(self):
+        """
+        Get the magnification for the highest-resolution level.
+
+        :return: magnification, width of a pixel in mm, height of a pixel in mm.
+        """
+        return {
+            'magnification': None,
+            'mm_x': None,
+            'mm_y': None
+        }
+
+    def getMagnificationForLevel(self, level=None):
+        """
+        Get the magnification at a particular level.
+
+        :param level: None to use the maximum level, otherwise the level to get
+            the magification factor of.
+        :return: magnification, width of a pixel in mm, height of a pixel in mm.
+        """
+        mag = self.getMagnification()
+
+        if level is not None and self.levels and level != self.levels - 1:
+            mag['scale'] = 2.0 ** (self.levels - 1 - level)
+            if mag['magnification']:
+                mag['magnification'] /= mag['scale']
+            if mag['mm_x'] and mag['mm_y']:
+                mag['mm_x'] *= mag['scale']
+                mag['mm_y'] *= mag['scale']
+        if self.levels:
+            mag['level'] = level if level is not None else self.levels - 1
+        return mag
```

Comment:
Then may be keep it as an internal method then. 

I feel it is nice to have only one way to do or get anything through the API.